### PR TITLE
[firebase] Update database schema and add queries for Transactions and VoucherRanges

### PR DESCRIPTION
**Pull request by selene-huang**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

## What's new in this PR
Firebase schema updates, new queries for voucher ranges, new queries for transactions, and updated queries for vouchers.

### Add VoucherRanges
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
This PR adds a new type called `VoucherRange` under `src/types/types.tsx`. In Firebase, these are stored in the `voucher-ranges` database. A `VoucherRange` object has the following fields:

- `startSerialNumber`: an integer representing the start of the range, inclusive
- `endSerialNumber`: an integer representing the end of the range, inclusive
- `type`: a string representing the color of vouchers in this range
- `maxValue`: an integer representing the maximum possible value of vouchers in this range in cents

Because app users should never have a reason to create or modify a `VoucherRange`, there are only two associated queries: `getAllVoucherRanges` and `getVoucherRange`.

**Note:** `getVoucherRange` is used to query the associated `voucherRange` for a given serial number. It will return a `VoucherRange` if it exists; otherwise, the function returns `null`.

### Update Vouchers Schema

The Firebase `vouchers` database schema has been updated. To reflect that, a `Voucher` object now has the following fields:

- `serialNumber`: an integer representing the voucher's unique serial number
   - **Note:** a string version of `serialNumber` is now used as the database's primary key for fast lookup
- `type`: a string representing the voucher color. Automatically calculated based on the `serialNumber`.
- `value`: an integer representing the voucher's value in cents
- `vendorUuid`: a string representing the current user's UUID

There are now only two associated get functions, `getVoucher` and `getVouchersByVendorUuid`. All update functions were removed, since users will not have the capability to update vouchers once they have been reviewed and submitted to the database.

Finally, `createVoucher` now includes three validation checks. To accommodate this, it returns a `createVoucherResult`, which is a json. If the query runs successfully, the field `ok` is set to `true` and `docId` contains the relevant doc id. Otherwise, `ok` is set to false, and `error` contains an error code from the `VoucherCreateError` enum.

- **InvalidSerialNumber**: `serialNumber` has no associated `voucherRange` and is invalid
- **SerialNumberAlreadyUsed**: `serialNumber` already exists in the `vouchers` database (i.e. it has already been used)
- **ValueExceededMaximum**: `value` is greater than the maximum possible value set in the associated `voucherRange`

**Full Change Log**:

- Changed the Firebase document id from autogenerated to the string version of the voucher's serial number
- Removed `VoucherStatus` enum under `src/types/types.tsx`
- Removed `status` field from `Voucher` type
- Removed `VoucherType` enum under `src/types/types.tsx`
- Added `VoucherCreateError` enum under `src/types/types.tsx`
- Added `VoucherCreateResult` type under `src/types/types.tsx`
- Added serial number validation handling under `src/database/queries.ts`
- Updated `createVoucher` to reflect schema changes - it now takes only `serialNumber`, `value`, and `vendorUuid` as parameters. Additional changes:
    - `type` is now automatically added based on the associated `VoucherRange`
    - Added validation check to ensure that `value` does not exceed the maximum possible value
- Removed `getAllVouchers` because there is no use case for it
- Removed `setVoucherVendorUuid` because there is no use case for it
- Removed `updateVoucher` because it is not used anywhere else

### Add Transactions
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

A new type called `Transaction` has been added under `src/types/types.tsx`. In Firebase, these are stored in the `transactions` database. A `Transaction` object has the following fields:

- `uuid`: a string that is automatically generated by Firebase
- `timestamp`: a Firebase `Timestamp`. This is automatically calculated when a transaction is created.
- `status`: a string selected from the `TransactionStatus` enum ('paid' or 'unpaid')
- `value`: an integer representing the sum of all values in `voucherSerialNumbers` *in cents*. This is automatically calculated when a transaction is created.
- `voucherSerialNumbers`: an array of integers, representing the serial numbers of all vouchers associated with this transaction
- `vendorUuid`: a string representing the current user's UUID

There are two associated get functions, `getTransaction` and `getTransactionsByVendorUuid`. Additionally, there is one update function, `setTransactionVoucherSerialNumbers`.

Finally, there is one new function that creates transactions, `createTransaction`. It uses the new type `TransactionCreate` in `src/types/types.tsx`, which ensures that a given `Transaction` is created with all the necessary user-defined values (`status`, `voucherSerialNumbers`, and `vendorUuid`).

### Miscellaneous Updates
- Added better documentation for `src/database/queries.ts`
- Changed all value amounts to cents (int) instead of dollars (float) to avoid incurring rounding errors from working with floats
- Updated all queries containing the field `vendor_uuid` to `vendorUuid`, to match the current database schema

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"

1. Import `VendorsListDemo.tsx` to `App.tsx` and verify that 'Create Voucher Out of Range', 'Create Duplicate Voucher', and 'Create Voucher with High Value' log the correct errors in the console.
2. Check that 'Create Voucher' correctly creates a new voucher with serial code **104**.
   - **Note**: This should only work once (because afterwards, you'll be creating a duplicate value), so if you would like to test repeatedly, either change the `serialNumber` to any other number in the set {102, 103, 105, 106, 107, 108, 109, 110}, or go into the `vouchers` database and delete the newly created document.
3. Verify that 'Create Transaction' correctly produces a new `Transaction` with value **101**.

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

This PR only adds or updates queries associated with Firebase. Local handling of data and its associated validation checks (e.g. checking that the user did not scan a voucher twice prior to submitting their vouchers to the `vouchers` database) will be added in future PRs.

There is also no handling for automatically fetching the current user and filling in their UUID for queries that require it. This will be introduced in future PRs as well.

## Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

Vouchers queries are refactored from PR #10.

[//]: # "This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation."
🥦 CC: @sauhardjain

### CODE REVIEW

File: VendorsListDemo.tsx [hunk 1]
```diff
@@ -9,21 +9,25 @@ import {
...
 } from './src/types/types';
 
 export default function VendorsListDemo() {
   const [vendors, setVendors] = useState<Vendor[]>([]);
   const [vouchers, setVouchers] = useState<Voucher[]>([]);
+  const [transactions, setTransactions] = useState<Transaction[]>([]);
   const [selectedVendor, setSelectedVendor] = useState<Vendor | null>(null);
 
-  /* fetch all vouchers on page load */
+  /* fetch all venders on page load */
```
Comment (wangannie) [inline VendorsListDemo.tsx]: nit: spell check -- "vendors"

File: package.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   "dependencies": {
+    "@calblueprint/prettier-config": "github:calblueprint/prettier-config",
```
Comment (wangannie) [inline package.json]: This change shouldn't be committed! Revert changes in this file.

Side note: was there an issue with your prettier config without this change?

File: src/database/queries.ts [hunk 3]
```diff
@@ -105,11 +150,50 @@ export const getVoucher = async (uuid: Uuid): Promise<Voucher> => {
...
+ *    `value`: monetary value of voucher in cents
+ *
+ *    `vendorUuid`: current user's Uuid
+ *
+ * Returns the doc id if the query executes successfully.
+ *
+ * Returns `'-1'` if the serial number is invalid,
+ * `'-2'` if the serial number has already been used,
+ * or `'-3'` if the voucher's value exceeds the maximum possible value.
+ */
```
Comment (wangannie) [inline src/database/queries.ts]: I like how you break down the errors into separate codes, but I think there's a cleaner way to do it with TypeScript:

I would create an enum of error codes:
```
enum VoucherCreateError {
  InvalidSerialNumber,
  SerialNumberAlreadyUsed,
  ValueExceedsMaximum,
}
```

and a new type `VoucherCreateResult` which is a union of the success and error cases:

```
type VoucherCreateResult =
  | { ok: true; docId: string }
  | { ok: false; error: VoucherCreateError };
```

The return type of `createVoucher` would then be `Promise<VoucherCreateResult>`.

When returning an error, you'd have something like 
```
return { ok: false, error: VoucherCreateError.InvalidSerialNumber };
```
or something like 
```
return { ok: true, docId };
```
in the success case.

File: src/database/queries.ts [hunk 4]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
 
 /**
- * Setter function to update a Voucher's VendorUuid
+ * Helper for calculating a Transaction's value.
+ *
+ * Returns the sum in cents of all vouchers in the given array.
  */
-export const setVoucherVendorUuid = async (uuid: Uuid, vendorUuid: Uuid) =>
-  updateVoucher({ uuid, vendorUuid });
+export const calculateValue = async (voucherArray: number[]) => {
```
Comment (wangannie) [inline src/database/queries.ts]: It's unclear here what `voucherArray` should be -- might be clearer if you rename `voucherArray` to `voucherSerialNumbers` or something like that.

Comment (wangannie) [reply src/database/queries.ts]: And rename `calculateValue` to make it clear exactly what this is doing, like `calculateVouchersTotalValue`?

File: src/database/queries.ts [hunk 5]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
+  try {
+    let value = 0;
```
Comment (wangannie) [inline src/database/queries.ts]: nit: rename to `totalValue`

File: src/types/types.tsx [hunk 6]
```diff
@@ -6,27 +8,40 @@ export type Vendor = {
   name: string;
 };
 
+export type VoucherRange = {
+  startSerialNum: number;
+  endSerialNum: number;
+  type: string;
```
Comment (wangannie) [inline src/types/types.tsx:14]: Consider using an enum for the type -- why was `VoucherType` removed?

File: src/types/types.tsx [hunk 7]
```diff
@@ -6,27 +8,40 @@ export type Vendor = {
...
-  GREEN = 'green',
-  ORANGE = 'orange',
-  PURPLE = 'purple',
-}
+export type Transaction = {
+  uuid: Uuid;
+  timestamp: Timestamp;
+  status: TransactionStatus;
+  value: number;
+  voucherArray: number[];
```
Comment (wangannie) [inline src/types/types.tsx]: Consider renaming to something to indicate that this is an array of voucher serial numbers bc otherwise this could be confused for an array of Voucher objects.

Something like `voucherSerialNumbers`?

Comment (wangannie) [reply src/types/types.tsx]: This applies to most places you use `voucherArray`

File: src/database/queries.ts [hunk 8]
```diff
@@ -75,27 +84,63 @@ export const getVendor = async (uuid: Uuid): Promise<Vendor> => {
...
 
-    return querySnapshots.docs.map(document => document.data() as Voucher);
+/**
+ * Query the `voucher-ranges` collection.
+ *
+ * Return a VoucherRange if the given serialNumber is in any voucher range.
+ *
+ * Return `-1` if no such VoucherRange is found.
+ */
+const getVoucherRange = async (serialNumber: number) => {
```
Comment (wangannie) [inline src/database/queries.ts]: You should specify a return type -- `Promise<VoucherRange | null>`

I also think it's better to just return `null` rather than `-1`, so it's easy to check whether a result has been found by truthiness without having to remember particular error codes.

File: src/database/queries.ts [hunk 9]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
+ */
+export const getTransactionsByVendorUuid = async (
   vendorUuid: Uuid,
-): Promise<Voucher[]> => {
+): Promise<Transaction[]> => {
   try {
     const dbQuery = query(
-      voucherCollection,
+      transactionCollection,
       where('vendor_uuid', '==', vendorUuid),
```
Comment (wangannie) [inline src/database/queries.ts]: On the vendors list demo, it wasn't actually displaying the vouchers by vendor, and I traced it to this: it should be `vendorUuid` so it matches how we have it in Firebase

Comment (wangannie) [reply src/database/queries.ts]: Same with `getVouchersByVendorUuid`

File: src/database/queries.ts [hunk 10]
```diff
@@ -105,11 +150,50 @@ export const getVoucher = async (uuid: Uuid): Promise<Voucher> => {
...
+      return '-2';
+    }
+
+    // check that value does not exceed the maximum
+    if (voucherRange.maxValue < voucher.value) {
+      return '-3';
+    }
+
+    await setDoc(docRef, voucher);
+    await updateDoc(docRef, { type: voucherRange.type });
```
Comment (wangannie) [inline src/database/queries.ts:197]: Looks like with this scheme, we're no longer using `uuid`s for each record, and instead we're using the serial number as a uuid. It seems to make sense given the assumption that serial numbers will always be unique, but I just wanted to flag it to make sure, since it strays from the existing convention. cc @sauhardjain

File: src/database/queries.ts [hunk 11]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
-const updateVoucher = async (voucher: Partial<Voucher>) => {
+export const getVouchersByVendorUuid = async (
+  vendorUuid: Uuid,
+): Promise<Voucher[]> => {
   try {
-    const docRef = doc(voucherCollection, voucher.uuid);
-    await updateDoc(docRef, voucher);
+    const dbQuery = query(
+      voucherCollection,
+      where('vendor_uuid', '==', vendorUuid),
```
Comment (wangannie) [inline src/database/queries.ts]: Change to `vendorUuid` to match our schema

File: src/database/queries.ts [hunk 12]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.warn('(calculateValue)', e);
+    throw e;
+  }
+};
 
 /**
- * Fetch all vouchers for a given vendor.
+ * Query to create a new transaction in Firebase.
```
Comment (wangannie) [inline src/database/queries.ts:268]: Very helpful comments! 👏🏻

File: src/database/queries.ts [hunk 13]
```diff
@@ -118,47 +202,147 @@ export const createVoucher = async (voucher: VoucherCreate): Promise<Uuid> => {
...
+    // eslint-disable-next-line no-console
+    console.warn('(updateTransaction)', e);
+    throw e;
+  }
+};
+
+/**
+ * Setter function to update a Transaction's status.
+ */
+export const setTransactionStatus = async (
```
Comment (wangannie) [inline src/database/queries.ts]: Will this actually need to be done in the client app? cc @sauhardjain

File: src/database/queries.ts (hunk 3 above)
Comment (wangannie) [reply src/database/queries.ts]: This also makes the error handling more intuitive, rather than having
```
const uuid = createVoucher ...`
```
where `uuid` could potentially be an error code

**Review by wangannie — CHANGES_REQUESTED** [review]
This is looking really good! Awesome job making this PR super clean, well-documented, and easy to review.

Most comments are cleanliness/renaming nits. The biggest piece I'd like to see changed is the error handling (using TypeScript magic rather than numeric or string error codes). Feel free to reach out if anything is unclear/you wanna walk through anything together!

File: src/database/queries.ts (hunk 3 above)
Comment (wangannie) [reply src/database/queries.ts]: Some docs the error handling method I proposed:
https://imhoff.blog/posts/using-results-in-typescript
https://journal.plain.com/posts/2022-10-04-error-handling-in-typescript-like-a-pro/
https://blog.logrocket.com/improve-error-handling-typescript-exhaustive-type-checking/

File: package.json (hunk 2 above)
Comment (selene-huang) [reply package.json]: oh yeah i was going through hell to get my prettier working, still not really sure what fixed it in the end

File: src/types/types.tsx (hunk 6 above)
Comment (sauhardjain) [reply src/types/types.tsx:14]: We decided to remove it because the VoucherType will never be entered in the app but only through the admin dashboard — in case the admins ever want to add a new VoucherType, we'd have to add another type in the enum but the string can just work as is.

File: src/database/queries.ts (hunk 10 above)
Comment (sauhardjain) [reply src/database/queries.ts:197]: Yeah, the purpose of doing this was to make sure look-ups for specific vouchers would be as quick as possible. 

@sauhardjain tagging myself to take care of this through Retool.

File: src/database/queries.ts (hunk 13 above)
Comment (sauhardjain) [reply src/database/queries.ts]: Good catch, we don't need this query in the client app.

File: package.json (hunk 2 above)
Comment (sauhardjain) [reply package.json]: Can you try discarding this change and see if prettier still works for you?

File: src/types/types.tsx (hunk 6 above)
Comment (selene-huang) [reply src/types/types.tsx:14]: yeh this was a database change. instead of users manually entering types and us just checking if the type exists, what happens in admins will create a new set of valid vouchers - say serial numbers 100-200 with color purple. when we validate a voucher's serial number, we perform a lookup on the voucher-ranges database to find the range it belongs to and automatically set the type. and like sauhard said, this has the added bonus of not needing to update any code if they want to add more types.

File: src/database/queries.ts (hunk 10 above)
Comment (selene-huang) [reply src/database/queries.ts:197]: we decided to make the change because the voucher database is expected to get very large, so using the serial number as the uuid should make lookups faster.

File: src/database/queries.ts [hunk 14]
```diff
@@ -103,28 +104,28 @@ export const getAllVoucherRanges = async (): Promise<VoucherRange[]> => {
 /**
  * Query the `voucher-ranges` collection.
  *
- * Return a VoucherRange if the given serialNumber is in any voucher range.
+ * Returns a voucherRange if the given serialNumber is in any voucher range.
```
Comment (wangannie) [inline src/database/queries.ts]: Nit: capitalizing VoucherRange (how you originally had it) would be good since it's the official type name!

File: src/database/queries.ts [hunk 15]
```diff
@@ -161,39 +162,40 @@ export const getVoucher = async (serialNumber: number): Promise<Voucher> => {
...
   try {
     const docId = voucher.serialNumber.toString();
     const docRef = doc(db, 'vouchers', docId);
 
     // check that serialNumber exists
     const voucherRange = await getVoucherRange(voucher.serialNumber);
-    if (voucherRange === -1) {
-      return '-1';
+    if (voucherRange === null) {
+      return { ok: false, error: VoucherCreateError.InvalidSerialNumber };
```
Comment (wangannie) [inline src/database/queries.ts:182]: Beautiful ✨

File: src/database/queries.ts [hunk 16]
```diff
@@ -278,7 +282,9 @@ export const createTransaction = async (
 ): Promise<Uuid> => {
   try {
     const docRef = await addDoc(transactionCollection, transaction);
-    const value = await calculateValue(transaction.voucherArray);
+    const value = await calculateTotalVouchersValue(
```
Comment (wangannie) [inline src/database/queries.ts:285]: YAY this is much more readable 🥇

**Review by wangannie — APPROVED** [review]
INCREDIBLE JOB SELENE!! You perfectly addressed all of my comments and this is looks so much cleaner ✨

**Review by sauhardjain — APPROVED** [review]
AMAZING STUFF @selene-huang. Everything looks great to me!

File: src/database/queries.ts (hunk 14 above)
Comment (sauhardjain) [reply src/database/queries.ts]: Should be good to merge after making this change!

File: src/types/types.tsx (hunk 6 above)
Comment (wangannie) [reply src/types/types.tsx:14]: Ah good thinking -- yes that makes sense to me! I'd just make a note of basically exactly what you just said in the PR description so it's documented clearly!
